doc/KUKSA.val_data_broker/authorization_alt_d.md

```diff
@@ -0,0 +1,358 @@
+    "client_id": "s6BhdRkqt3",
+    "scope": "read:Vehicle actuate:Vehicle"
+}
+```
+
+#### Data Structure
+
+| Claim     | Description                                          |
+|-----------|------------------------------------------------------|
+| iss       | REQUIRED - as defined in Section 4.1.1 of [RFC7519]. |
+| exp       | REQUIRED - as defined in Section 4.1.4 of [RFC7519]. |
+| aud       | REQUIRED - as defined in Section 4.1.3 of [RFC7519]. See Section 3 for indications on how an authorization server should determine the value of "aud" depending on the request. |
+| sub       | REQUIRED - as defined in Section 4.1.2 of [RFC7519]. In cases of access tokens obtained through grants where a resource owner is involved, such as the authorization code grant, the value of "sub" SHOULD correspond to the subject identifier of the resource owner. In cases of access tokens obtained through grants where no resource owner is involved, such as the client credentials grant, the value of "sub" SHOULD correspond to an identifier the authorization server uses to indicate the client application. See Section 5 for more details on this scenario. Also, see Section 6 for a discussion about how different choices in assigning "sub" values can impact privacy. |
+| client_id | REQUIRED - as defined in Section 4.3 of [RFC8693].   |
+| iat       | REQUIRED - as defined in Section 4.1.6 of [RFC7519]. This claim identifies the time at which the JWT access token was issued. |
+| jti       | REQUIRED - as defined in Section 4.1.7 of [RFC7519]. |
+
+#### Authorization claims
+By specifying a scope when requesting an access token from the authorization server,
+the resource owner can limit what a client (e.g. a third party app) is allowed to do.
+
+> If an authorization request includes a scope parameter, the corresponding issued
+> JWT access token SHOULD include a "scope" claim as defined in Section 4.2 of [RFC8693].
+
+> All the individual scope strings in the "scope" claim MUST have meaning for the
+> resources indicated in the "aud" claim. See Section 5 for more considerations about
+>  the relationship between scope strings and resources indicated by the "aud" claim.
+
+| Claim     | Description                                          |
+|-----------|------------------------------------------------------|
+| scope     | The value of the scope claim is a JSON string containing a space-separated list of scopes associated with the token, in the format described in Section 3.3 of [RFC6749] |
+
+### Scope
+
+Access to VSS entries / signals are limited by specifying a scope in the access token provided to
+KUKSA.VAL.
+
+#### Format
+
+The scopes used in KUKSA.VAL are "prefix scopes" that consist of a VSS path, prefixed
+with an action.
+
+| Scope                 | Description                 |
+|-----------------------|-----------------------------|
+|`<ACTION>:<PATH>`      | Allow ACTION for PATH       |
+|`!<ACTION>:<PATH>`     | Deny ACTION for PATH        |
```

Comment:
I still have a strong dislike for the negative scopes/"deny rules", I feel even more strongly so when using the scope

 - It just feels unnatural to me. I did try to find any wording in any  RFC that this is forbidden, but could not. None of the posted examples did this though
   - The Linus Torvalds argument here is maybe "this is so stupid, that nobody before us ever though about it"
 - It adds complication, i.e. you need add special handling and explanantion, such as the precedence below, that makes it less obvious for users when having situations like.  read:Vehicle.Body.Trunk !read:Vehicle:Body
 - If we ever extend/add scopes that include others, i.e. read can read "all", while there a specific scopes for certain (meta)data such as current values/QoS information , etc. it gets more complex
 - I think it is really not needed. The only _potential_(!) problem it solves are to large scope claims, because you want to include too many scattered datapoints that a negative list is smaller than a positive. We might (for a long time)never hit this. If we do there might be any number of other solutions for that (i.e. the tagging discussed below may be one such thing)
 - Leaving it out now, would not preclude a path of adding it later, as it is just another scope. But I really think we shouldn't.

Reply:
I fully agree with the sentiment.

> Leaving it out now, would not preclude a path of adding it later, as it is just another scope. But I really think we shouldn't.

Agreed. I'll move it to another "future possible extension" instead.

Reply:
One thing that is rather different in our use case, in comparison to most examples that you can find out there, is that we rely exclusively on limiting access by what is provided in the access token.

I would say the common use case is that the subject usually issuing an access token has a number of access restrictions to begin with. And these have to be checked in addition to the scope by the resource server.

So if we are not adding additional ways to limit the access rights for a group of users, one use case for these "deny" scopes would be that the authorization server automatically adds them to all access tokens for a certain group of users.

But that's perhaps a misuse of the feature, and there are better ways to solve it.